src/Middleware/HttpLogging/src/W3CLoggingMiddleware.cs

```diff
@@ -98,7 +98,9 @@ public async Task Invoke(HttpContext context)
 
             if (options.LoggingFields.HasFlag(W3CLoggingFields.Time))
             {
-                shouldLog |= AddToList(elements, _timeIndex, now.ToString("HH:mm:ss", CultureInfo.InvariantCulture));
+                // Start time will be logged to second precision, but we store it to millisecond precision so we can use it
+                // to calculate time-taken later.
+                shouldLog |= AddToList(elements, _timeIndex, now.ToString("HH:mm:ss.fff", CultureInfo.InvariantCulture));
```

Comment:
Can we calculate and set the final value for `_timeTakenIndex` in this middleware? That way we wouldn't have to parse this date and call substring on this later in `W3CLogger.Format(string[])`. We could also use `Environment.TickCount` to avoid misreporting the elapsed time if the system clock changes.

Reply:
Good idea

Reply:
Looks like Environment.TickCount isn't as precise, it's in whole milliseconds so most request/response pairs log a time-taken of 0

Reply:
Wasn't this counting whole milliseconds before? Even if we continue using DateTime to calculate elapsed time, we should move where we calculate elapsed time. StopWatch might also be an option, but I'd want to be careful to avoid too many allocations.

Reply:
> StopWatch might also be an option, but I'd want to be careful to avoid too many allocations.

`ValueStopWatch` could be used to avoid allocation.